### DISCUSSION TITLE
Writing to the bytes()

### DISCUSSION THREAD

**Discussion by telenkor** [jim-easterbrook/python-exiv2#13] [Q&A] [answered]
I don't understand. There is an object that, in essence, is a file, only in RAM. Why can't I update his Exif? Although a check in the code shows that the original Exif has been erased, and only a single new key remains. But when saving from "preview.data" to disk, everything remains the same.
I tried wrapping it in BytesIO, but it still doesn’t update :cry:

``` python
import exiv2
import rawpy

with rawpy.imread("/Users/test/Desktop/_DVC0754.NEF") as raw:
        preview = raw.extract_thumb()
        if preview.format == rawpy.ThumbFormat.JPEG:

            # Here we could save this JPEG to disk, but we need to change its Exif first.
            # with open("/Volumes/Disk/output.jpg", 'wb') as file:
            #         file.write(preview.data)

            # Create new key and value "Exif.Image.Software"
            exifData_out = exiv2.ExifData()
            key = exiv2.ExifKey("Exif.Image.Software")
            v = exiv2.Value.create(exiv2.TypeId.asciiString)
            v.read("Some Text XXX»)
            exifData_out.add(key, v)

            # Trying to write to the original Exif.
            image = exiv2.ImageFactory.open(preview.data)
            image.setExifData(exifData_out)
            image.writeMetadata()

            # Here we only check that "image" has taken on the new value "Exif.Image.Software"
            # All metadata has been erased, leaving only the new value. Great!
            image.readMetadata()
            exifData = image.exifData()
            i = exifData.begin()
            end = exifData.end()
            while i != end:
                print('{:44s} {:04x} {:9s} {:3d} {:s}'.format(
                    i.key(), i.tag(), i.typeName(), i.count(), i.toString()))
                next(i)

            # But file is saved with the old value
            with open("/Volumes/Disk/output.jpg", 'wb') as file:
                    file.write(preview.data)
```
Anticipating the question that Exiv2 is also able to extract previews from a raw image, I will answer this way: in my process there is not only extraction of previews, but also conversion from raw. Therefore, for Exiv2 I leave the work exclusively with Metadata.

**Comment by jim-easterbrook** [comment]
When you pass `preview.data` to libexiv2 it (probably) copies the data, or as much of it as it needs. No subsequent operation in the C++ library libexiv2 can affect the contents of the Python object `preview.data`.

You probably need to call `image.io()` to get the modified data. See
https://exiv2.org/doc/classExiv2_1_1Image.html#a62345361501892f42c6bbc86514bf5c6

`io.mmap()` is used to get a Python buffer to access the data without copying.
```
io = image.io()
with io.mmap() as image_data:
    with open("/Volumes/Disk/output.jpg", 'wb') as file:
        file.write(image_data)
io.munmap()
io.close()
```

**Reply by jim-easterbrook** [reply, marked as answer]
PS Instead of this:
```
            key = exiv2.ExifKey("Exif.Image.Software")
            v = exiv2.Value.create(exiv2.TypeId.asciiString)
            v.read("Some Text XXX»)
            exifData_out.add(key, v)
```
you could have this:
```
            exifData_out["Exif.Image.Software"] = "Some Text XXX"
```

**Comment by telenkor** [comment]
Thanks for the tip.

`
with io.mmap() as image_data:
`
 —> ((<class 'TypeError'>, 'exiv2.preview.byte_buffer' object does not support the context manager protocol, <traceback object at 0x111e8c080>))

It worked with this.
```python
            io = image.io()
            with open("/Volumes/Disk/output.jpg", 'wb') as file:
                file.write(io.mmap())
            io.munmap()
            io.close()
```

**Reply by jim-easterbrook** [reply]
Ah yes, I'd copied my answer from somewhere where I was using `with memoryview(io.mmap()) as image_data:`. The `memoryview` has the context manager.

**Reply by jim-easterbrook** [reply]
I've added a context manager to `Exiv2::BasicIO` objects to get easy read-only access to data, so in the next release you will be able to replace
```
            io = image.io()
            io.open()
            with open("/Volumes/Disk/output.jpg", 'wb') as file:
                file.write(io.mmap())
            io.munmap()
            io.close()
```
with
```
            with image.io() as data:
                with open("/Volumes/Disk/output.jpg", 'wb') as file:
                    file.write(data)
```

I need to expand USAGE.rst to include more detail about this.

**Reply by jim-easterbrook** [reply]
I decided the context manager was not very intuitive, so I've gone back to providing a Python buffer interface to the `Exiv2::BasicIO` data (using `mmap()`  etc internally). But now the buffer interface can be writeable, allowing direct modification of the data. Saving data to a file gets even easier:
```
            with open("/Volumes/Disk/output.jpg", 'wb') as file:
                file.write(image.io())
```
